File: haskell/PointInHull.hs
{-# LANGUAGE BangPatterns #-}

-- Point-in-convex-hull (2D) benchmark
module Main where

import qualified Data.ByteString.Char8 as BS
import Data.Int (Int64)
import qualified Data.Vector.Unboxed.Mutable as MV
import Numeric (showFFloat)
import System.Clock (Clock (Monotonic), getTime, toNanoSecs)
import System.Environment (getArgs)

readSignedInt :: BS.ByteString -> (Int64, BS.ByteString)
readSignedInt !bs =
    let !bs' = BS.dropWhile (\c -> c /= '-' && (c < '0' || c > '9')) bs
     in case BS.readInt bs' of
            Just (!n, !rest) -> (fromIntegral n, rest)
            Nothing -> error "parse error"

main :: IO ()
main = do
    args <- getArgs
    case args of
        [file] -> run file
        _ -> putStrLn "Usage: point_in_hull <polygon_file>"

run :: FilePath -> IO ()
run file = do
    -- ── Read ─────────────────────────────────────────────────────────────
    t0 <- getTime Monotonic
    contents <- BS.readFile file

    let (!np64, r0) = readSignedInt contents
        (!nq64, r1) = readSignedInt r0
        !np = fromIntegral np64 :: Int
        !nq = fromIntegral nq64 :: Int

    px <- MV.new np :: IO (MV.IOVector Int64)
    py <- MV.new np :: IO (MV.IOVector Int64)
    let readPoly !rem !i
            | i >= np = return rem
            | otherwise = do
                let (!x, r2) = readSignedInt rem
                    (!y, r3) = readSignedInt r2
                MV.write px i x
                MV.write py i y
                readPoly r3 (i + 1)
    r2 <- readPoly r1 0

    qx <- MV.new nq :: IO (MV.IOVector Int64)
    qy <- MV.new nq :: IO (MV.IOVector Int64)
    let readQueries !rem !i
            | i >= nq = return ()
            | otherwise = do
                let (!x, r3) = readSignedInt rem
                    (!y, r4) = readSignedInt r3
                MV.write qx i x
                MV.write qy i y
                readQueries r4 (i + 1)
    readQueries r2 0

    t1 <- getTime Monotonic
    let readMs = fromIntegral (toNanoSecs t1 - toNanoSecs t0) / 1e6 :: Double

    -- ── Compute ──────────────────────────────────────────────────────────
    t2 <- getTime Monotonic

    let checkQuery !q !count
            | q >= nq = return count
            | otherwise = do
                !x <- MV.read qx q
                !y <- MV.read qy q
                let checkEdge !i
                        | i >= np = return True
                        | otherwise = do
                            let !j = if i + 1 >= np then 0 else i + 1
                            !pxi <- MV.read px i
                            !pyi <- MV.read py i
                            !pxj <- MV.read px j
                            !pyj <- MV.read py j
                            let !cross = (pxj - pxi) * (y - pyi) - (pyj - pyi) * (x - pxi)
                            if cross < 0
                                then return False
                                else checkEdge (i + 1)
                !isIn <- checkEdge 0
                checkQuery (q + 1) (if isIn then count + 1 else count)
    !inside <- checkQuery 0 (0 :: Int)

    t3 <- getTime Monotonic
    let computeMs = fromIntegral (toNanoSecs t3 - toNanoSecs t2) / 1e6 :: Double

    putStrLn $
        "read="
            ++ showFFloat (Just 1) readMs ""
            ++ "ms compute="
            ++ showFFloat (Just 1) computeMs ""
            ++ "ms inside="
            ++ show inside
            ++ " total="
            ++ show nq

File: rust/src/bin/point_in_hull.rs
// Point-in-convex-hull (2D) benchmark
use std::env;
use std::fs;
use std::time::Instant;

#[inline(always)]
fn parse_i64(bytes: &[u8]) -> i64 {
    if bytes.first() == Some(&b'-') {
        let mut n: i64 = 0;
        for &b in &bytes[1..] {
            n = n * 10 + (b - b'0') as i64;
        }
        -n
    } else {
        let mut n: i64 = 0;
        for &b in bytes {
            n = n * 10 + (b - b'0') as i64;
        }
        n
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <polygon_file>", args[0]);
        std::process::exit(1);
    }

    // ── Read ─────────────────────────────────────────────────────────────────
    let t0 = Instant::now();

    let bytes = fs::read(&args[1]).unwrap();
    let mut iter = bytes
        .split(|&b| b == b' ' || b == b'\n' || b == b'\r' || b == b'\t')
        .filter(|s| !s.is_empty());

    let np: usize = parse_i64(iter.next().unwrap()) as usize;
    let nq: usize = parse_i64(iter.next().unwrap()) as usize;

    let mut px = vec![0i64; np];
    let mut py = vec![0i64; np];
    for i in 0..np {
        px[i] = parse_i64(iter.next().unwrap());
        py[i] = parse_i64(iter.next().unwrap());
    }

    let mut qx = vec![0i64; nq];
    let mut qy = vec![0i64; nq];
    for i in 0..nq {
        qx[i] = parse_i64(iter.next().unwrap());
        qy[i] = parse_i64(iter.next().unwrap());
    }

    let read_ms = t0.elapsed().as_secs_f64() * 1000.0;

    // ── Compute ──────────────────────────────────────────────────────────────
    let t1 = Instant::now();

    let mut inside = 0u32;
    for q in 0..nq {
        let x = qx[q];
        let y = qy[q];
        let mut is_in = true;
        for i in 0..np {
            let j = if i + 1 < np { i + 1 } else { 0 };
            let cross = (px[j] - px[i]) * (y - py[i]) - (py[j] - py[i]) * (x - px[i]);
            if cross < 0 {
                is_in = false;
            }
        }
        if is_in {
            inside += 1;
        }
    }

    let compute_ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!(
        "read={:.1}ms compute={:.1}ms inside={} total={}",
        read_ms, compute_ms, inside, nq
    );
}
